runtime/lua/vim/lsp/util.lua

```diff
@@ -771,72 +771,97 @@ function M.text_document_completion_list_to_complete_items(result, prefix)
   return matches
 end
 
---- Like vim.fn.bufwinid except it works across tabpages.
-local function bufwinid(bufnr)
-  for _, win in ipairs(api.nvim_list_wins()) do
-    if api.nvim_win_get_buf(win) == bufnr then
-      return win
+local function path_components(path)
+  return vim.split(vim.fs.normalize(path, { expand_env = false }), '/', { plain = true })
+end
+
+local function path_under_prefix(path, prefix)
+  for i, c in ipairs(prefix) do
+    if c ~= path[i] then
+      return false
     end
   end
+  return true
 end
 
---- Get list of buffers for a directory
-local function get_dir_bufs(path)
-  path = path:gsub('([^%w])', '%%%1')
+--- Get list of buffers whose filename matches given path prefix
+---@return integer[]
+local function get_bufs_with_prefix(prefix)
+  prefix = path_components(prefix)
   local buffers = {}
   for _, v in ipairs(vim.api.nvim_list_bufs()) do
-    local bufname = vim.api.nvim_buf_get_name(v):gsub('buffer://', '')
-    if bufname:find(path) then
+    local path = path_components(vim.api.nvim_buf_get_name(v))
+    if path_under_prefix(path, prefix) then
       table.insert(buffers, v)
     end
   end
   return buffers
 end
 
+local function escape_gsub_pat(s)
+  local result, _ = s:gsub('(%p)', '%%%1')
+  return result
+end
+
+local function escape_gsub_repl(s)
```

Comment:
does there can't use `vim.fn.escape` ?

Reply:
lua-pattern and gsub {repl} are escaped with `%`, but escape() uses `\`

Reply:
Does `vim.psub()` not work?

Reply:
forget `vim.pesc` then.